upstream(node): Porting node team related upstream changes in range mainnet-1.32.2 to mainnet-1.33.3

muXxer opened the pull request:
# Description of change

This PR contains a batch of node-team related upstream changes. The changes have been individually reviewed before squash-merging. None of them modify protocol parameters, so a new protocol version is not necessary.

This PR should be rebased on top of develop without squashing as it already contains squashed PRs.

## Links to any relevant issues

List of included PRs:

* #6179 
* #6181 
* #6182 
* #6183 
* #6334 
* #6184 
* #6623 
* #6390
* #6414 
* #6440
* #6786 
* #6791 
* #6817 

List of included upstream commits:

* https://github.com/MystenLabs/sui/commit/3a68b7dbf1b39a5f6c722517ff9b76623cc07573
* https://github.com/MystenLabs/sui/commit/07f8b22d5615ad2bcb355d32b02317e024987da2
* https://github.com/MystenLabs/sui/commit/346775946303e0e432685a7aebbe4ae108ed5e88
* https://github.com/MystenLabs/sui/commit/a22b71b70d6330f33c165198c194e4b716bc953b
* https://github.com/MystenLabs/sui/commit/070a2c38134a4ff4afaca150218289a3c8927603
* https://github.com/MystenLabs/sui/commit/6a0adb5de7f3fb87d3aa23364f200b542e3e8b87
* https://github.com/MystenLabs/sui/commit/a8e969ad338c6c0f26a66eb893eaf18432f08e3c
* https://github.com/MystenLabs/sui/commit/82bf44e58df032e5a8dd5e452c51a1aa6959f15e
* https://github.com/MystenLabs/sui/commit/fa7419bfcfeffaa816e3232c5d6af5db67dc7ca8
* https://github.com/MystenLabs/sui/commit/0e78656655529c130e6b680e0962ff6204306713
* https://github.com/MystenLabs/sui/commit/e1abe0963b579d69471ce40bd465d63fb80f0f41
* https://github.com/MystenLabs/sui/commit/262390b27ca7befe9bd3a537aae6ebef9f569348
* https://github.com/MystenLabs/sui/commit/42dbb6253959e1deda9afe4f34c0d717b9d67ccc

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration tests)
- [ ] Patch-specific tests (correctness, functionality coverage)

### Release Notes

- [ ] Protocol:
- [x] Nodes (Validators and Full nodes):
  * Migrate users of iota_metrics Histogram to prometheus Histogram
  * Rename metric names:
    * last_created_checkpoint_age_ms => last_created_checkpoint_age
    * last_certified_checkpoint_age_ms => last_certified_checkpoint_age
    * checkpoint_creation_latency_ms => checkpoint_creation_latency
    * checkpoint_contents_age_ms => checkpoint_contents_age
    * last_executed_checkpoint_age_ms => last_executed_checkpoint_age
    * checkpoint_summary_age_ms => checkpoint_summary_age
    * checkpoint_prepare_latency_us => checkpoint_prepare_latency
    * checkpoint_exec_latency_us => checkpoint_exec_latency
    * handle_soft_bundle_certificates_count => validator_service_handle_soft_bundle_certificates_count
    * handle_soft_bundle_certificates_size_bytes => validator_service_handle_soft_bundle_certificates_size_bytes
  * Add metric for active time of monitored futures (monitored_future_active_duration_ns)
  * Add metric for number of rocksdb files and level 0 files (rocksdb_total_num_files, rocksdb_num_level0_files)
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

bingyanglin committed:
node-upstream-change([v1.32.2, v1.33.3)): Emit error from shared counter creation failure (#6179)

# Description of change

Port
https://github.com/MystenLabs/sui/commit/3a68b7dbf1b39a5f6c722517ff9b76623cc07573

From [Sui's commit
description](https://github.com/MystenLabs/sui/commit/3a68b7dbf1b39a5f6c722517ff9b76623cc07573):
Otherwise we don't see the error from quorum driver.

## Links to any relevant issues

Part of #3990 

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Ran the local network with `RUST_LOG=info cargo run --release --bin iota
start --force-regenesis --with-faucet`

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items
that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have checked that new and existing unit tests pass locally with
my changes

bingyanglin committed:
node-upstream-change([v1.32.2, v1.33.3)): Caching resolver for Tonic (#6181)

# Description of change

Port
https://github.com/MystenLabs/sui/commit/07f8b22d5615ad2bcb355d32b02317e024987da2

From [Sui's commit
description](https://github.com/MystenLabs/sui/commit/07f8b22d5615ad2bcb355d32b02317e024987da2):
On some platforms (GKE) the local resolver does not seem to cache
aggressively, which results in high variance for requests

## Links to any relevant issues

Part of #3990 

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Ran the local network with `RUST_LOG=info cargo run --release --bin iota
start --force-regenesis --with-faucet`

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items
that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have checked that new and existing unit tests pass locally with
my changes

---------

Co-authored-by: jkrvivian <jkrvivian@gmail.com>

bingyanglin committed:
node-upstream-change([v1.32.2, v1.33.3)): Put back HandleConsensusOutput scope (#6182)

# Description of change

Port
https://github.com/MystenLabs/sui/commit/346775946303e0e432685a7aebbe4ae108ed5e88

From [Sui's commit
description](https://github.com/MystenLabs/sui/commit/346775946303e0e432685a7aebbe4ae108ed5e88):
Put back HandleConsensusOutput scope

Note that in the file `crates/sui-core/src/consensus_handler.rs`, you
won't see the following code snippet in our codebase. The following is
from a pretty old PR: [Use a single db batch per narwhal commit in
consensus task.
#10360](https://github.com/MystenLabs/sui/pull/10360/files), and we do
need it.

```rust
        // This code no longer supports old protocol versions.
         assert!(self
             .epoch_store
          ...
 ```

## Links to any relevant issues

Part of #3990 

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Ran the local network with `RUST_LOG=info cargo run --release --bin iota start --force-regenesis --with-faucet`

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

bingyanglin committed:
node-upstream-change([v1.32.2, v1.33.3)): add monitored scopes to CheckpointExecutor (#6183)

# Description of change

Port
https://github.com/MystenLabs/sui/commit/a22b71b70d6330f33c165198c194e4b716bc953b

From [Sui's commit
description](https://github.com/MystenLabs/sui/commit/a22b71b70d6330f33c165198c194e4b716bc953b):
add monitored scopes to CheckpointExecutor

## Links to any relevant issues

Part of #3990 

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Ran the local network with `RUST_LOG=info cargo run --release --bin iota
start --force-regenesis --with-faucet`

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items
that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have checked that new and existing unit tests pass locally with
my changes

bingyanglin committed:
node-upstream-change([v1.32.2, v1.33.3)): Improvements to thread stall monitor (#6334)

# Description of change

Port
https://github.com/MystenLabs/sui/commit/070a2c38134a4ff4afaca150218289a3c8927603

From [Sui's commit
description](https://github.com/MystenLabs/sui/commit/070a2c38134a4ff4afaca150218289a3c8927603):
Use a watchdog thread for monitoring thread stalls.

This allows us to detect a thread stall while it is still occurring,
rather than only after the fact.

## Links to any relevant issues

Part of #3990 

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Ran the local network with `RUST_LOG=info cargo run --release --bin iota
start --force-regenesis --with-faucet`

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items
that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have checked that new and existing unit tests pass locally with
my changes

bingyanglin committed:
node-upstream-change([v1.32.2, v1.33.3)): Delete sst files in key range in rocksdb (#6184)

# Description of change

Port
https://github.com/MystenLabs/sui/commit/6a0adb5de7f3fb87d3aa23364f200b542e3e8b87

From [Sui's commit
description](https://github.com/MystenLabs/sui/commit/6a0adb5de7f3fb87d3aa23364f200b542e3e8b87):
This PR adds a rocksdb endpoint to delete .sst files in key range in
rocksdb which is useful to prune data (in certain scenarios) without
compaction.

## Links to any relevant issues

Part of #3990 

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Ran the local network with `RUST_LOG=info cargo run --release --bin iota
start --force-regenesis --with-faucet`

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items
that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have checked that new and existing unit tests pass locally with
my changes

bingyanglin committed:
node-upstream-change([v1.32.2, v1.33.3)): Add caching for HttpKVStore (#6623)

# Description of change

- Port [Sui's commit
#18995](https://github.com/MystenLabs/sui/commit/a8e969ad338c6c0f26a66eb893eaf18432f08e3c)
- Add caching for HttpKVStore


## Links to any relevant issues

Part of #3990 

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Ran the local network with `RUST_LOG=info cargo run --release --bin iota
start --force-regenesis --with-faucet`

**Note**: There is protocol config change and the snapshot should be
updated afterwards (not in this PR) to make the testing pass.

semenov-vladyslav committed:
Migrate users of iota_metrics Histogram to prometheus Histogram (#6390)

iota_metrics variant is difficult to use in Grafana because its
precomputation of percentiles makes it impossible to aggregate streams
from multiple hosts in a statistically valid way.

This keeps a few of the mysten_metrics Histogram versions around to help
with the transition of exsiting users.

Port of
https://github.com/MystenLabs/sui/commit/82bf44e58df032e5a8dd5e452c51a1aa6959f15e

Part of #3990

- Upstream commit port

semenov-vladyslav committed:
Refactor ConsensusTransactionKind (#6414)

# Description of change

Port of
https://github.com/MystenLabs/sui/commit/fa7419bfcfeffaa816e3232c5d6af5db67dc7ca8

PR contains some changes that are partially reverted in subsequent
upstream commits, notably:
- RandomnessStateUpdate was removed in this PR and later reverted back,
not it's in the upstream HEAD.
- AuthorityCapabilitiesV2 are used in the upstream which were not
ported.
- renames UserTransaction to CertifiedTransaction, but UserTransaction
is reintroduced later again.

Should be merged before PR for
https://github.com/MystenLabs/sui/commit/c1384f97e67b805f5a915742892206a65f120b54.

## Links to any relevant issues

Part of #3990 

## Type of change

- Upstream commit port

semenov-vladyslav committed:
Add metric for active time of monitored futures (#6440)

# Description of change

Port of
https://github.com/MystenLabs/sui/commit/0e78656655529c130e6b680e0962ff6204306713

## Links to any relevant issues

Part of #3990 

## Type of change

- Upstream commit port

gokhan-simsek-iota committed:
node-upstream-change([v1.32.2, v1.33.3)): Add a metric to track the number of managed files per column family in RocksDb (#6786)

# Description of change
Port
[MystenLabs/sui@e1abe09](https://github.com/MystenLabs/sui/commit/e1abe0963b579d69471ce40bd465d63fb80f0f41)

Add a metric to track the number of managed files per column family in
RocksDb (https://github.com/MystenLabs/sui/pull/19214). We don't add
more detailed level to control the cardinality of the metrics.


## Links to any relevant issues

Part of https://github.com/iotaledger/iota/issues/3990

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

`cargo r --bin iota-node -- --config-path fullnode.yaml`

`RUST_LOG=info cargo r --bin iota start --force-regenesis --with-faucet`

gokhan-simsek-iota committed:
node-upstream-change([v1.32.2, v1.33.3)): [rocksdb] add metric for num level 0 files (#6791)

## Description

Port
[MystenLabs/sui@262390b](https://github.com/MystenLabs/sui/commit/262390b27ca7befe9bd3a537aae6ebef9f569348)

Too many level 0 files can be a trigger to write stall & stop as well

## Links to any relevant issues

Part of https://github.com/iotaledger/iota/issues/3990

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested
```
cargo r --bin iota-node -- --config-path fullnode.yaml
```
```
RUST_LOG=info cargo r --bin iota start --force-regenesis --with-faucet
```

jkrvivian committed:
feat: disable write stall on fullnodes perpetual db (#6817)

# Description of change

Port
https://github.com/MystenLabs/sui/commit/42dbb6253959e1deda9afe4f34c0d717b9d67ccc

From Sui's commit description:
For fullnodes that do not prune the `perpetual` DB, especially
`transactions` and `effects` cfs, they can run into write stalls
occasionally that makes the fullnode non operational. Since fullnodes
have to accept all writes from checkpoints, throttling writes do not
seem to make much sense.
Write stalling on validators is left enabled.

## Links to any relevant issues

Part of #3990 

## Type of change
- Enhancement 

---------

Co-authored-by: muXxer <git@muxxer.de>

semenov-vladyslav commented:
Some iota-graphql-e2e-tests fail due to new `"content-type": "application/graphql-response+json"`. [This comment](https://github.com/graphql/graphql-over-http/issues/249#issuecomment-1503736569) mentions this content type.

muXxer commented:
> Some iota-graphql-e2e-tests fail due to new `"content-type": "application/graphql-response+json"`. [This comment](https://github.com/graphql/graphql-over-http/issues/249#issuecomment-1503736569) mentions this content type.

should be fixed with the latest rebase

semenov-vladyslav committed:
feat(node): remove IotaHistogram metrics (#7043)

# Description of change

Remove the remaining deprecated IotaHistogram metrics. There are a total
of 6 metrics using IotaHistogram to report various time intervals in
milliseconds; all of them already have counterpart metrics using
prometheus Histogram and report time in seconds with fractions. The
removed metrics are not used in community dashboard.

## Links to any relevant issues

Fixes #6766.

## Type of change

- Enhancement (a non-breaking change which removes deprecated
functionality)

### Release Notes

- [ ] Protocol:
- [x] Nodes (Validators and Full nodes): Removed deprecated metrics.
(last_created_checkpoint_age_ms, last_certified_checkpoint_age_ms,
checkpoint_creation_latency_ms, checkpoint_contents_age_ms,
last_executed_checkpoint_age_ms, checkpoint_summary_age_ms)
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

---------

Co-authored-by: muXxer <git@muxxer.de>
